File: sample-super-rx-ai/src/main/java/com/mateuszkoslacz/moviper/rxsample/viper/contract/ListingContract.java
package com.mateuszkoslacz.moviper.rxsample.viper.contract;

import com.hannesdorfmann.mosby.mvp.MvpView;
import com.mateuszkoslacz.moviper.iface.interactor.ViperRxInteractor;
import com.mateuszkoslacz.moviper.iface.routing.ViperRxRouting;
import com.mateuszkoslacz.moviper.rxsample.viper.entity.User;

import java.util.List;

import rx.Observable;
import rx.subjects.Subject;

public interface ListingContract {

    interface View extends MvpView {
        // Defines what methods the Presenter can invoke on the View
        // In most cases there will be manipulating ui and displaying data or errors.
        // In Super Rx version it also provides getters for Observables emmiting user click events.

        Subject<User, User> getUserClicks();

        void setUserList(List<User> userList);

        void showError(Throwable throwable);

        void showLoading();

        void showContent();
    }

    interface Interactor extends ViperRxInteractor {
        // Defines what methods the Presenter can invoke on the Interactor.
        // In most cases there will be data saving and querying.
        // It's just a marker interface.

        Observable<List<User>> getUserList();
    }

    interface Routing extends ViperRxRouting {
        // Defines what methods the Presenter can invoke on the Routing.
        // In most cases there will be starting another activities, services and using system
        // framework, ie. scheduling alarms or sending broadcasts.
        // In the case of a fragment being the view, there also will be manipulating
        // the root Activity, ie. switching fragments.

        void startUserDetailsActivity(User user);
    }
}

File: sample-super-rx-ai-kotlin/src/main/java/com/mateuszkoslacz/moviper/rxsample/viper/contract/ListingContract.kt
package com.mateuszkoslacz.moviper.rxsample.viper.contract

import android.app.Activity
import com.hannesdorfmann.mosby.mvp.MvpView
import com.mateuszkoslacz.moviper.iface.interactor.ViperRxInteractor
import com.mateuszkoslacz.moviper.iface.routing.ViperRxRouting
import com.mateuszkoslacz.moviper.rxsample.viper.entity.User
import rx.Observable

interface ListingContract {

    interface View : MvpView {
        val userClicks: Observable<User>

        fun setUserList(userList: List<User>)

        fun showError(throwable: Throwable)

        fun showLoading()

        fun showContent()
    }

    interface Interactor : ViperRxInteractor {
        val userList: Observable<List<User>>
    }

    interface Routing : ViperRxRouting<Activity> {
        fun startUserDetailsActivity(user: User)
    }
}
